src/typegenTypeHelpers.ts

```diff
@@ -255,3 +260,221 @@ export type NeedsResolver<TypeName extends string, FieldName extends string> = H
+
+/**
+ * Returns a union of all the abstract type names where TypeName is used
+ *
+ * @example
+ * union D = A | B
+ * union E = A
+ * AbstractTypeNames<'A'> // 'D' | 'E'
+ */
+export type AbstractTypeNames<TypeName extends string> = ConditionalKeys<
+  GetGen<'abstractResolveReturn'>,
+  TypeName
+>
+
+/**
+ * Returns whether all the abstract type names where TypeName is used have implemented `resolveType`
+ */
+export type ResolveTypeImplementedInAllAbstractTypes<TypeName extends string> = AbstractTypeNames<
+  TypeName
+> extends GetGen<'resolveTypeImplemented'>
+  ? true
+  : false
+
+/**
+ * Returns whether all the members of an abstract type have implemented `isTypeOf`
+ */
+export type IsTypeOfImplementedInAllMembers<AbstractTypeName extends string> = GetGen2<
+  'abstractResolveReturn',
+  AbstractTypeName
+> extends GetGen<'isTypeOfObjectNames'>
+  ? true
+  : false
+
+export type IsTypeOfHandler<TypeName extends string> = (
+  source: PossibleTypes<TypeName>, // typed as never if TypeName is not a member of any abstract type
+  context: GetGen<'context'>,
+  info: GraphQLResolveInfo
+) => MaybePromise<boolean>
+
+/**
+ * Conditionally returns the `isTypeOf` field
+ */
+export type IsTypeOf<TypeName extends string> = GetGen3<
+  'features',
+  'abstractTypes',
+  'isTypeOf',
+  false
+> extends false
+  ? {}
+  : ResolveTypeImplementedInAllAbstractTypes<TypeName> extends true
+  ? {
+      isTypeOf?: IsTypeOfHandler<TypeName>
+    }
+  : // Make isTypeOf optional as soon as __typename is enabled
+  GetGen3<'features', 'abstractTypes', '__typename', false> extends true
+  ? {
+      isTypeOf?: IsTypeOfHandler<TypeName>
+    }
+  : { isTypeOf: IsTypeOfHandler<TypeName> }
+
+/**
+ * Conditionally returns the `resolveType` field
+ */
+export type ResolveType<TypeName extends string> = GetGen3<
+  'features',
+  'abstractTypes',
+  'resolveType',
+  false
+> extends false
+  ? {} // remove field altogether is feature is not enabled
+  : IsTypeOfImplementedInAllMembers<TypeName> extends true
+  ? {
+      /**
+       * Optionally provide a custom type resolver function. If one is not provided,
+       * the default implementation will call `isTypeOf` on each implementing
+       * Object type.
+       */
+      resolveType?: AbstractTypeResolver<TypeName>
+    }
+  : // Make resolveType optional as soon as __typename is enabled
+  GetGen3<'features', 'abstractTypes', '__typename', false> extends true
+  ? {
+      /**
+       * Optionally provide a custom type resolver function. If one is not provided,
+       * the default implementation will call `isTypeOf` on each implementing
+       * Object type.
+       */
+      resolveType?: AbstractTypeResolver<TypeName>
+    }
+  : {
+      /**
+       * Optionally provide a custom type resolver function. If one is not provided,
+       * the default implementation will call `isTypeOf` on each implementing
+       * Object type.
+       */
+      resolveType: AbstractTypeResolver<TypeName>
+    }
```

Comment:
Nice set of building block funcs. All of this could go under an `AbstractTypes` namespace (see later comment).